### PR TITLE
fix: handle OSError importing cairosvg to avoid build crash because of missing system dependencies

### PR DESCRIPTION
**Pull request by Guts**
Closes: #6817

### CODE REVIEW

**Comment by squidfunk** [comment]
Thanks for the PR. However, we're not considering merging it.

The reason is that if we wouldn't throw the error, it might go silent in CI builds, with users wondering why social cards are not building. We're already experiencing high demand in support for setting up Cairo, and this would lead to even more issues why social cards are not building, as users might forget to install the dependencies, so the build will succeed. Thus, we cannot succeed the build if the dependencies are not available.

A better solution is to only enable the social cards plugin only in CI using an [environment variable](https://www.mkdocs.org/user-guide/configuration/#environment-variables), as the `CI` environment variable is already defined in most CI build environments, e.g., GitHub Actions and GitLab:

``` yaml
plugins:
  - social:
      enabled: !ENV CI
```

This will disable generation of social cards locally (which makes no sense anyway, as you cannot post local links to Twitter or other social media platforms) and keep them when built within CI. We always need to keep the economics of maintaining the project in sight when considering PRs for merging. I hope for your understanding. Thank you.

**Comment by Guts** [comment]
I get your rationale.  
However, there are 2 commits in my PR precisely because I wanted to separate :

1. the [first](ee8a6480d60d5ab0090b3ef2f1c9a8ad8c7c935a) prevents mkdocs from crashing (cf the issue) but let the actual plugin's behavior (exiting the build with an exception)
1. the [second](30b1d6cee5344e7f4354f4eb8b77232310996ca1) suggests avoiding the error and continuing the build; which will fail if the user has opted for strict mode, which is how the Mkdocs ecosystem works.

According to your argument, I can understand that the second point is debatable and why not assume that end users are not aware of strict mode and don't read the logs (I have the same experience on my own projects). But the former should be integrated to reduce support queries following a mkdocs build crash, shouldn't it?

For the record, I'm already using the environment variable to activate/deactivate the plugin, but I haven't figured out how to activate it only on a Linux environment or depending on the presence of Cairo dependencies.

**Comment by squidfunk** [comment]
Okay! Let's fix the `OSError` then – could you remove the second commit or open a new PR?

**Comment by kamilkrzyskow** [comment]
I see that @Guts has reacted, but no change 🤔So was "you" in "could you remove the second commit or open a new PR?" directed at me specifically or both of us 😅 
Will you adjust the PR @Guts or am I free to do it myself? I think I'm eligible to make changes with "Maintainers are allowed to edit this pull request. ", because I have push rights 🤔

**Comment by Guts** [comment]
Sorry, I did not have time past days and not much either in coming next.

So feel free to make the changes!

**Comment by squidfunk** [comment]
It was targeted towards @Guts, the OP of the PR, but I don't mind you making the changes, @kamilkrzyskow ☺️

**Comment by kamilkrzyskow** [comment]
I made the errors a bit more explicit:
```bash
ERROR   -  Required dependencies of "social" plugin not found.
Install with: pip install "mkdocs-material[imaging]"
No module named 'PIL'
No module named 'cairosvg'
```

```bash
ERROR   -  "cairosvg" Python module is installed, but it crashed with:
no library called "cairo-2" was found
no library called "cairo" was found
no library called "libcairo-2" was found
cannot load library 'libcairo.so.2': error 0x7e.  Additionally, ctypes.util.find_library() did not manage to locate a library called 'libcairo.so.2'
cannot load library 'libcairo.2.dylib': error 0x7e.  Additionally, ctypes.util.find_library() did not manage to locate a library called 'libcairo.2.dylib'
cannot load library 'libcairo-2.dll': error 0x7e.  Additionally, ctypes.util.find_library() did not manage to locate a library called 'libcairo-2.dll'
---
Did you install the external CairoSVG library?
Did you add the binaries path to your system PATH?
Did you restart the Terminal to assure the PATH was reloaded?
```

File: material/plugins/social/plugin.py
Comment (squidfunk) [inline material/plugins/social/plugin.py:57]: So I'm not sure I'm for using this import error concatenation pattern, but I'm short on time now thinking of a better approach, so I'm happy to test it out. Is that a common practice?

File: material/plugins/social/plugin.py
Comment (squidfunk) [inline material/plugins/social/plugin.py:99]: What if we create a troubleshooting section in our image processing documentation and link to it? I think this might be a better approach because we have essentially infinite space as opposed to a terminal, it is easier to maintain and format, and we can cater to more OSs.

We then would only include the link here.

**Review by squidfunk — CHANGES_REQUESTED** [review]
Thanks for your work! I'd say we move out the troubleshooting info to our documentation.

File: material/plugins/social/plugin.py
Comment (kamilkrzyskow) [reply material/plugins/social/plugin.py:57]: > Is that a common practice?

Converting the error to string is a valid practice, but I don't know if it's common. The idea with gathering the errors to print them out later came from MkDocs <-> Plugin relationship. Errors that are local to the plugin should be properly moved upstream to the event flow and I wanted to be more specific about the message that says "dependencies not installed", so I needed to store the errors from the config load to report them later on.
If you feel it's weird I can just revert it :v:

File: material/plugins/social/plugin.py
Comment (kamilkrzyskow) [reply material/plugins/social/plugin.py:99]: OK, I will write up a small troubleshooting guide and put a link. Should I use the t.ly link shortener again?

File: material/plugins/social/plugin.py
Comment (squidfunk) [reply material/plugins/social/plugin.py:57]: Let's just keep it like that for now and revisit it later, all good! I just wanted to ask.

File: material/plugins/social/plugin.py
Comment (squidfunk) [reply material/plugins/social/plugin.py:99]: Yes, [t.ly](https://t.ly/) is the one we're using because it seems to be the most sensible to use right now.

File: material/plugins/social/plugin.py
Comment (kamilkrzyskow) [reply material/plugins/social/plugin.py:57]: Funnily enough, when researching things for the troubleshooting guide I found out that the cairocffi (dependency of cairosvg) also gathers the exceptions and the result we see are a mix of formatted strings and errors joined together 😅 

So the commonality increased :v:
https://github.com/Kozea/cairocffi/blob/f1984d644bbc462ef0ec33b97782cf05733d7b53/cairocffi/__init__.py#L39-L40

**Comment by kamilkrzyskow** [comment]
The current errors look like this:
```bash
ERROR   -  Required dependencies of "social" plugin not found:
           - ModuleNotFoundError("No module named 'PIL'")
           - ModuleNotFoundError("No module named 'cairosvg'")

           --> Install with: pip install "mkdocs-material[imaging]"
```
```bash
ERROR   -  "cairosvg" Python module is installed, but it crashed with:
           no library called "cairo-2" was found
           no library called "cairo" was found
           no library called "libcairo-2" was found
           cannot load library 'libcairo.so.2': error 0x7e.  Additionally, ctypes.util.find_library() did
           not manage to locate a library called 'libcairo.so.2'
           cannot load library 'libcairo.2.dylib': error 0x7e.  Additionally, ctypes.util.find_library()
           did not manage to locate a library called 'libcairo.2.dylib'
           cannot load library 'libcairo-2.dll': error 0x7e.  Additionally, ctypes.util.find_library() did
           not manage to locate a library called 'libcairo-2.dll'

           --> Check out the troubleshooting guide: https://t.ly/dlwQ7
```

The troubleshooting guide first and foremost recommends to restart the Terminal windows to reload PATH and stuff.
Then goes into a bit more detail, as best as I could write it out. I didn't check the macOS script yet (with GitHub Actions would be my only option). Perhaps I should provide some simple command instead of the script 🤔 
In Windows Powershell there is 
```powershell
$env:Path -split ';'
```
and it will display a list, just like the one in the Python script, but without the caironames etc.

I'll change the state of the PR to draft, since I wait for feedback :v:

**Comment by kamilkrzyskow** [comment]
Made a test repository and run the scripts on each system:
https://github.com/kamilkrzyskow/test-macos-actions/actions/runs/8103418921

Everything works as expected, however it kind of sucks I can't create a simple Linux script for easier troubleshooting.
Unless I would create it specifically for Ubuntu / GitHub Actions and tell other disctro users to deal with it on their own.

---

I also think I might've misunderstood and the troubleshooting guide should've been at the bottom of the page, not inside the cairo section? 🤔

File: docs/plugins/requirements/image-processing.md
Comment (squidfunk) [inline docs/plugins/requirements/image-processing.md:96]: Why the anchor name? I'd say we'll keep it as troubleshooting to make room for more to add later + keep the anchor links stable. If we feel that we need more granularity, we can move troubleshooting one way up and add further headlines.

File: docs/plugins/requirements/image-processing.md
Comment (squidfunk) [inline docs/plugins/requirements/image-processing.md:138]: Just an idea: we _could_ create a `etc/debug` folder in our repository and move the script there. That way we can version it + we could, if necessary, use Snippets to integrate it. In the future, we might also somehow move this logic into the social plugin (or better: into our own utilities folder, so we can use it in other plugins as well) and/or bundle the script the theme (not now, let's keep it simple for now).

What do you think? I'm not sure if this makes sense, just wanted to bounce ideas.

File: docs/plugins/requirements/image-processing.md
Comment (kamilkrzyskow) [reply docs/plugins/requirements/image-processing.md:96]: OK, so I misunderstood, I'll move the Troubleshooting to the bottom of the page as another H2 tag, a sibling of Dependencies, and change the link to `#troubleshooting` 👌

File: docs/plugins/requirements/image-processing.md
Comment (kamilkrzyskow) [reply docs/plugins/requirements/image-processing.md:138]: There is already the https://github.com/squidfunk/mkdocs-material/tree/master/includes directory. We could move the scripts there into the `etc/debug` structure, and later include them with Snippets, as you said. 

I'm still not sure about the Linux scripts, because GitHub Actions finds the library with https://www.unix.com/man-page/Linux/1/ld/ but I'm not sure if that will always be the case with the cairo library or if it depends on how someone installs it 😑 

> social plugin (or better: into our own utilities folder, so we can use it in other plugins as well) and/or bundle the script the theme (not now, let's keep it simple for now).

The scripts I provided are supposed to display the paths that are checked or the way they're checked for the user that wants to troubleshoot them. They are somewhat generalized, since they run with the `ctypes.util` built-in and not cairoffi directly, but I'm doubtful they would be useful as utils in the plugins. 

Another idea I had, was to ask you to create a mkdocs-material/troubleshooting repository and link to it from that troubleshooting guide, and there would be CI similar to my own repository, which people could copy to execute the actions and see the scripts in action, but this idea is a bit "out there" doesn't really focus on the matter at hand of getting it to work on the users PC, and requires "GitHub Actions proficiency" of the user.

Comment (squidfunk) [reply docs/plugins/requirements/image-processing.md:138]: Yes, let's move it into `include/debug` and use Snippets for now.

> Another idea I had, was to ask you to create a mkdocs-material/troubleshooting repository and link to it from that troubleshooting guide, and there would be CI similar to my own repository, which people could copy to execute the actions and see the scripts in action, but this idea is a bit "out there" doesn't really focus on the matter at hand of getting it to work on the users PC, and requires "GitHub Actions proficiency" of the user.

Interesting idea, I'll discuss it with the core team next time we speak ☺️

**Comment by kamilkrzyskow** [comment]
Thanks to Snippets the Markdown became more readable, and allowed for a bit longer Linux script.
I've focused on Ubuntu and Manjaro, as the other BSD or SunOS Linux variants aren't that common.
Moved the structure and adjusted the t.ly link in the scripts.

The updated scripts can be seen running here:
https://github.com/kamilkrzyskow/test-macos-actions/actions/runs/8159277013

**Comment by kamilkrzyskow** [comment]
Did another force push to span out the text to 80 characters in some places,

**Comment by Guts** [comment]
Impressive and really valuable work 👏!

Thanks @kamilkrzyskow!

**Comment by kamilkrzyskow** [comment]
The last thing I'm unsure about, as this wasn't in the change request before, is that I did set CAPITAL-cased global variables, and from the rest of the code base it seems that global `log` variables are lowercase, so there is no distinction between global and local case?

**Review by squidfunk — APPROVED** [review]
LGTM! Thanks for putting the time in, great work!

One final remark when viewing the documentation though: I think we should not use Snippets, as the scripts are very long and that adds a lot of unnecessary noise to the documentation. This was my fault, recommending it but I also did not know the scripts will get so long. The scripts do not _explain_ anything, and are more or less intended to be "just run" by the user helping with troubleshooting, so I would say we just link to the `raw.githubusercontent` versions of them, ideally directly triggering a download if that is possible.

**Comment by kamilkrzyskow** [comment]
> The scripts do not explain anything, and are more or less intended to be "just run" by the user helping with troubleshooting

Yes, that was my intention from the start, as I believe adding a lot of comments / print statements, would just make it longer, and a non-techy person would still not read it before running it. There is only one print at the end, saying if it found the path. 

> so I would say we just link to the raw.githubusercontent versions of them, ideally directly triggering a download if that is possible.

I will look into a command to download the script and pipe them to Python. I've seen some PowerShell scripts do that, but without Python. I think the same should be possible with a `curl` command on Unix.

**Comment by squidfunk** [comment]
> I will look into a command to download the script and pipe them to Python. I've seen some PowerShell scripts do that, but without Python. I think the same should be possible with a curl command on Unix.

Perfect. Also link to the script that is downloaded via curl, so users can inspect it. Anything else would be fishy.

**Comment by kamilkrzyskow** [comment]
Done and a test here: https://github.com/kamilkrzyskow/test-macos-actions/actions/runs/8227806730
Ultimately, I also changed the UPPERCASE variables to lowercase, as I stumbled upon this in the search.py:
https://github.com/squidfunk/mkdocs-material/blob/c8bae9b5b4b46f0032a9f2dddb30adda68158c78/src/plugins/search/plugin.py#L33-L36
Not entirely the same, but still consistent with the lowercase `log`, so all variables are lowercase in the codebase, I hope I didn't mess it up now, as the previous review accepted the uppercased variables...

**Review by squidfunk — APPROVED** [review]
LGTM. I need to find a time slot to be able to merge this through to Insiders, but since the social plugin is entirely different there, I expect it to take some time. Please bear with me.

**Comment by kamilkrzyskow** [comment]
I can also make a PR towards the Insiders branch, if you want to to have the dependency logic the same there, but since the plugins have branched out into two different versions, I don't think they have to merge the change at the same time.

**Comment by kamilkrzyskow** [comment]
I had a look at the source code in Insiders, and I see that the dependency check and related error was moved out of `on_config` to `_generate` which is invoked concurrently after `on_page_markdown` 😵💫
I'm not sure why the change was made, the comment claims Pillow to be the minimal requirement, but there are no safety checks done for cairosvg, so it's assumed it's always installed anyway.

I would move out the safety logic back to `on_config` like the community version of the plugin, as I don't see advantages to throwing an error only after the concurrency jobs are invoked 🤔

**Comment by squidfunk** [comment]
Yes, that's why I said I need a little more time to investigate. I think I put it there because of a better editing experience when crafting social card layouts. I'm quite sure I put it in `on_config` first and moved it out because of several oddities, but I can't remember entirely – sorry, too many changes happened since then and my comment is not clear enough (that's why I stress to write good comments explaining design decisions so often).

We need to test it thoroughly before making any change to it, i.e. check how the social plugin behaves in all of those cases (building, serving, cairo error, pillow not available). I would like to work on this myself. Again, please give me some time.

**Comment by vedranmiletic** [comment]
Thanks for this, social plugin now works again inside pipenv venv on FreeBSD.

**Pull request merged**
